ovn-org/ovn-kubernetes
# Adds port security for external OVS bridge for OVN MAC address

If the MAC changes on the interface used in OVS and the OVS bridge, OVN may still retain the old MAC on its GR. This can cause OVN and the host to both respond with different MACs for the same IP address, leading to network disruption. A common scenario for this is a server with a bond. When the node is rebooted, the bond MAC may change depending on which slave comes up first. When this happens, ovnkube-node starts up and is unable to sync its informers and proceed to full start up, due to disruption with kapi.

Note, this can be avoided by setting the MAC specifically for the bond to always be the mac of the first slave.

This PR adds port security in a sense to the breth0/br-ex bridge. It drops any traffic sent from the OVN patch port that has a MAC other than the one we expect by modifying the flows that OFManager installs. 

Additionally, at start up there can be only a normal flow on the switch. This is common on node reboot. When this happens in the scenario above, ovnkube-node cannot get to the part of the code that starts the OFManager, because it is blocked waiting for the informers to sync. For this purpose, a new bootstrap function is added which will install these basic security flows if only a single flow exists in the bridge at start up.

I looked at moving OFManager to start earlier, but it is tightly coupled with gateway initialization and kapi objects, so rather than introducing a new regression there I created a simple bootstrap function. The bootstrap flows are wiped when OFManager starts and syncs its new flows (which include the security flows anyway).

How to test:
1. Launch kind
2. docker exec -it ovn-worker /bin/bash
3. crictl exec -it <ovs pod> /bin/bash
4. Inside ovs, create flows.txt with the content: priority=0, table=0, actions=output:NORMAL
5. ovs-ofctl -O openflow13 replace-flows breth0 flows.txt
6. quickly crictl stop the ovnkube-node pod
7. ovs-ofctl -O openflow13 replace-flows breth0 flows.txt
8. while true; do echo "##########" >> output.txt && ovs-ofctl -O openflow13 dump-flows breth0 >> output.txt && sleep .5; done
9. check output.txt to see that the bootstrap detected the normal flow correctly and configured the right flows:

```
##########
OFPST_FLOW reply (OF1.3) (xid=0x2): 
 cookie=0x0, duration=686.264s, table=0, n_packets=2142, n_bytes=941544, priority=0 actions=NORMAL
##########
OFPST_FLOW reply (OF1.3) (xid=0x2):
 cookie=0xdeff105, duration=0.203s, table=0, n_packets=0, n_bytes=0, priority=10,in_port=2,dl_src=02:42:ac:12:00:03 actions=NORMAL
 cookie=0xdeff105, duration=0.203s, table=0, n_packets=0, n_bytes=0, priority=9,in_port=2 actions=drop
 cookie=0x0, duration=686.768s, table=0, n_packets=2167, n_bytes=944776, priority=0 actions=NORMAL
##########
OFPST_FLOW reply (OF1.3) (xid=0x2): 
 cookie=0xdeff105, duration=0.429s, table=0, n_packets=0, n_bytes=0, priority=500,ip,in_port=2,nw_src=172.18.0.3,nw_dst=169.254.169.2 actions=ct(commit,table=4,zone=64001,nat(dst=172.18.0.3))
 cookie=0xdeff105, duration=0.429s, table=0, n_packets=0, n_bytes=0, priority=500,ip,in_port=LOCAL,nw_dst=169.254.169.1 actions=ct(table=5,zone=64002,nat)
 cookie=0xdeff105, duration=0.429s, table=0, n_packets=0, n_bytes=0, priority=500,ip,in_port=LOCAL,nw_dst=10.96.0.0/16 actions=ct(commit,table=2,zone=64001,nat(src=169.254.169.2))
 cookie=0xdeff105, duration=0.429s, table=0, n_packets=0, n_bytes=0, priority=105,ip,in_port=2,nw_dst=10.96.0.0/16 actions=drop
 cookie=0xdeff105, duration=0.429s, table=0, n_packets=0, n_bytes=0, priority=500,ip,in_port=2,nw_src=10.96.0.0/16,nw_dst=169.254.169.2 actions=ct(table=3,zone=64001,nat)
 cookie=0xdeff105, duration=0.429s, table=0, n_packets=0, n_bytes=0, priority=205,udp,in_port=1,dl_dst=02:42:ac:12:00:03,tp_dst=6081 actions=LOCAL
 cookie=0xdeff105, duration=0.429s, table=0, n_packets=0, n_bytes=0, priority=200,udp,in_port=1,tp_dst=6081 actions=NORMAL
 cookie=0xdeff105, duration=0.429s, table=0, n_packets=0, n_bytes=0, priority=200,udp,in_port=LOCAL,tp_dst=6081 actions=output:1

```
